pkg/storage/stores/local/downloads.go

```diff
@@ -45,37 +38,39 @@ func (s *Shipper) checkStorageForUpdates(ctx context.Context, period string, fc
 			toDownload = append(toDownload, object)
 		}
 	}
+	fc.RUnlock()
 
-	for uploader := range fc.files {
+	err = fc.ForEach(func(uploader string, df *downloadedFile) error {
 		if _, isOK := listedUploaders[uploader]; !isOK {
 			toDelete = append(toDelete, uploader)
 		}
-	}
+		return nil
+	})
 
 	return
 }
 
-// syncFilesForPeriod downloads updated and new files from for given period from all the uploaders and removes deleted ones
-func (s *Shipper) syncFilesForPeriod(ctx context.Context, period string, fc *filesCollection) error {
-	level.Debug(util.Logger).Log("msg", fmt.Sprintf("syncing files for period %s", period))
-
-	fc.RLock()
-	toDownload, toDelete, err := s.checkStorageForUpdates(ctx, period, fc)
-	fc.RUnlock()
+// Sync downloads updated and new files from for given period from all the uploaders and removes deleted ones
+func (fc *FilesCollection) Sync(ctx context.Context) error {
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf("syncing files for period %s", fc.period))
 
+	toDownload, toDelete, err := fc.checkStorageForUpdates(ctx)
 	if err != nil {
 		return err
 	}
 
 	for _, storageObject := range toDownload {
-		err = s.downloadFile(ctx, period, storageObject, fc)
+		err = fc.downloadFile(ctx, storageObject)
 		if err != nil {
 			return err
 		}
 	}
 
+	fc.Lock()
+	defer fc.Unlock()
+
 	for _, uploader := range toDelete {
-		err := s.deleteFileFromCache(period, uploader, fc)
+		err := fc.cleanupFile(uploader)
```

Comment:
I think this would be better expressed as `fc.CleanupFiles(toDelete...)`, which could then be refactored like
```
func (fc *FilesCollection) CleanupAllFiles(files ...string)
```

This cleanup function already takes the lock in the exact same way, so we'd be able to increase code reuse and be more idiomatic.